amaranth/vendor/_gatemate.py

```diff
@@ -0,0 +1,120 @@
+from abc import abstractmethod
+from amaranth import *
+from amaranth.build import *
+from amaranth.lib.cdc import ResetSynchronizer
+
+__all__ = ["GateMatePlatform"]
+
+class GateMatePlatform(TemplatedPlatform):
+    """
+    .. rubric:: GateMate FPGA toolchain
+
+    Required tools:
+        * ``yosys``
+        * ``p_r``
+
+    The environment is populated by setting the ``CC_TOOL`` environment variable to point to the toolchain directory.
+    """
+
+    device = property(abstractmethod(lambda: None))
+    package = property(abstractmethod(lambda: None))
+
+    toolchain = "GateMate"
+
+    required_tools = [
+        "yosys",
+        "p_r",
+    ]
+
+    file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.ccf": r"""
+            # {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                Net "{{port_name}}" Loc = "{{pin_name}}"
+                {%- set constraints = [
+                    "SCHMITT_TRIGGER", "PULLUP", "PULLDOWN", "KEEPER", "SLEW", "DRIVE", 
+                    "DELAY_OBF", "DELAY_IBF", "FF_IBF", "FF_OBF", "LVDS_BOOST", "LVDS_TERM"
+                ] -%}
+                {%- for constraint in constraints if constraint in attrs -%}
+                    | {{constraint}}={{attrs[constraint]}}
+                {%- endfor -%};
+            {% endfor %}
+        """,
+        "{{name}}.sdc": r"""
+            # {{autogenerated}}
+            {% for net_signal, port_signal, frequency in platform.iter_clock_constraints() -%}
+                {% if port_signal is not none -%}
+                    create_clock -period {{1000000000/frequency}} [get_ports {{port_signal.name|tcl_quote}}]
+                {% endif %}
+            {% endfor %}
+        """,
+    }
+
+    command_templates = [
+        r"""
+        mkdir -p log;
+        mkdir -p net;
+        {{invoke_tool("yosys")}}
+            -ql log/synth.log -p 'read -sv {{name}}.v; synth_gatemate -top {{name}} -nomx8 -vlog net/{{name}}_synth.v'
+        """,
+        r"""
+
+        {{invoke_tool("p_r")}}
+            -v
+            -i net/{{name}}_synth.v
+            -o {{name}}
+            -ccf {{name}}.ccf
+            -cCP > log/impl.log
+        """,
+    ]
+
+    # Common logic
+
+    @property
+    def default_clk_constraint(self):
+        if self.default_clk == "sys_clk0":
+            return Clock(self.osc_freq / self.osc_div)
+        return super().default_clk_constraint
+
+    def add_clock_constraint(self, clock, frequency):
+        super().add_clock_constraint(clock, frequency)
+        clock.attrs["keep"] = "TRUE"
```

Comment:
Does the GateMate toolchain (the `p_r` tool) recognize the `keep` attribute? (Yosys does, and ignores the value.)

Reply:
What is the keep attribute?

Reply:
It's documented in the Yosys README.

Reply:
The `p_r` tool does some optimization that doesn't seem to be linked to yosys's keep attribute. Here are all the optimization parameters for the `p_r`.
![image](https://github.com/user-attachments/assets/1d1d6ebe-c49f-46e1-bdfd-f70bd22a0483)

Reply:
Which PDF is this from? I can't find it on the website.

Reply:
You can find the all their documentations [here](https://colognechip.com/programmable-logic/gatemate/gatemate-documentation/).
The one in question here is [GateMate CCGM1A1/A2 Datasheet (May 2024)](https://colognechip.com/docs/ds1001-gatemate1-datasheet-latest.pdf) which is the Chip that is used on the Olimex and I think also on the Starter board.

Reply:
I didn't realize the datasheet would have the P&R options, that's fairly unusual... I'm reviewing it now.

Reply:
Do they really not have any way to place constraints, not even to get an STA failure if you can't meet them? That is really unusual, but I guess that's what the migration to nextpnr is for.

You should delete the SDC file entirely then.

Reply:
Deleted the SDC file part.